src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -381,80 +390,93 @@ entails SolverOptions{..} constraint context hints =
-            -- If both arguments are applications of the same type constructor
-            -- (e.g. @data D a b = D a@ in the constraint
-            -- @Coercible (D a b) (D a' b')@), infer the roles of the type
-            -- constructor's arguments and generate wanted constraints
-            -- appropriately (e.g. here @a@ is representational and @b@ is
-            -- phantom, yielding @Coercible a a'@).
-            let tyRoles = lookupRoles env aTyName
-                k role ax bx = case role of
-                  Nominal
-                    -- If we had first-class equality constraints, we'd just
-                    -- emit one of the form @(a ~ b)@ here and let the solver
-                    -- recurse. Since we don't we must compare the types at
-                    -- this point and fail if they don't match. This likely
-                    -- means there are cases we should be able to handle that
-                    -- we currently can't, but is at least sound.
-                    | ax == bx ->
-                        Just []
-                    | otherwise ->
-                        Nothing
-                  Representational ->
-                    Just [Constraint nullSourceAnn C.Coercible [] [ax, bx] Nothing]
-                  Phantom ->
-                    Just []
-            fmap concat $ sequence $ zipWith3 k tyRoles axs bxs
-        | (TypeConstructor _ tyName, _, xs) <- unapplyTypes t
-        , not $ null xs
-        , Just (tvs, wrappedTy, _) <- lookupNewtypeConstructor env tyName -> do
-            -- If the first argument is a newtype applied to some other types
-            -- (e.g. @newtype T a = T a@ in @Coercible (T X) b@), look up the
-            -- type of its wrapped field and yield a new wanted constraint in
-            -- terms of that type with the type arguments substituted in (e.g.
-            -- @Coercible (T[X/a]) b = Coercible X b@ in the example).
-            let wrappedTySub = replaceAllTypeVars (zip tvs xs) wrappedTy
-            pure [Constraint nullSourceAnn C.Coercible [] [wrappedTySub, b] Nothing]
-      _ ->
+    coercibleWanteds :: Environment -> SourceType -> SourceType -> m (Maybe [SourceConstraint])
+    coercibleWanteds env a b
+      | (TypeConstructor _ aTyName, _, axs) <- unapplyTypes a
+      , (TypeConstructor _ bTyName, _, bxs) <- unapplyTypes b
+      , Just (aTyKind, _) <- M.lookup aTyName $ types env
+      , Just (bTyKind, _) <- M.lookup bTyName $ types env
+      , (aks, kind) <- unapplyKinds aTyKind
+      , (bks, _) <- unapplyKinds bTyKind
+      , length axs < length aks
+      , length bxs < length bks = do
+          -- If both arguments have kind @k1 -> k2@ (e.g. @data D a b = D a@
+          -- in the constraint @Coercible (D a) (D a')@), yield a new wanted
+          -- constraint in terms of the types saturated with the same variables
+          -- (e.g. @Coercible (D a t0) (D a' t0)@ in the exemple).
+          tys <- traverse freshTypeWithKind aks
```

Comment:
Will this work if the type constructors in use in `a` and `b` have different numbers of arguments? I'm wondering if this should instead be something like `replicate (length aks - length axs) freshTypeWithKind`, and then get rid of the `drop` on the following lines. For example, if we have, say
```
MkX :: X -> Type -> Type
MkYY :: Y -> Y -> Type -> Type
SomeX :: X
SomeY :: Y
```
and the two arguments `a` and `b` we're dealing with are `MkX SomeX` and `MkYY SomeY SomeY`, then I think we will have
```
axs = [ SomeX ]
bxs = [ SomeY, SomeY ]
aks = [ X, Type ]
bks = [ Y, Y, Type ]
```
In this case, `tys` will have the same number of elements as `aks`, i.e. 2, and so we'll have
```
drop (length axs) tys = [ t0 ]
drop (length bxs) tys = []
```
so we end up recursing with 
```
a = MkX SomeX t0
b = MkYY SomeY Some Y
```
which then leads to a kind unification error, even though the user might not have written anything ill-kinded? Let me know if this makes any sense.

Reply:
Very well spotted! Thank you for taking the time to write such a detailed example 🙇

Reply:
Sure! I was mainly doing it for my own benefit initially, to try to check I had understood it properly (and because this was too much to keep in my head at once) haha

Reply:
Would you mind adding a test which would catch this too, when you get a chance?

Reply:
I added tests/purs/failing/CoercibleHigherKindedData.purs and a case inspired by https://github.com/purescript/purescript/pull/3893#issuecomment-687634696 to tests/purs/passing/Coercible.purs.